File: razer-sdk/src/main/java/net/booone/razersdk/sdk/ProgressiveRzKeySelector.java
package net.booone.razersdk.sdk;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public final class ProgressiveRzKeySelector {
    private final int length;
    private final List<Set<RzKey>> parts;
    private int currentStep = 0;

    public ProgressiveRzKeySelector(List<Set<RzKey>> parts) {
        this(parts, 1);
    }

    public ProgressiveRzKeySelector(List<Set<RzKey>> parts, int length) {
        if (parts == null || parts.isEmpty()) {
            throw new IllegalArgumentException("There must be at least one part");
        }

        if (length <= 0) {
            throw new IllegalArgumentException("Length must be a positive number");
        } else if (length > parts.size()) {
            throw new IllegalArgumentException("Length cannot be larger than number of parts");
        }

        this.parts = parts;
        this.length = length;
    }

    public Set<RzKey> getNextPart() {
        int span = length - 1;
        if (currentStep == parts.size() + span) {
            currentStep = 0;
        }
        Set<RzKey> returnedPart;
        if (currentStep - span <= 0) {
            returnedPart = joinSets(0, currentStep);
        } else if (currentStep >= parts.size() - 1) {
            returnedPart = joinSets(currentStep - span, parts.size() - 1);
        } else {
            returnedPart = joinSets(currentStep - span, currentStep);
        }
        currentStep++;
        return returnedPart;
    }

    public int getTotalSteps() {
        return parts.size() + length - 1;
    }

    private Set<RzKey> joinSets(int startIndex, int endIndex) {
        return parts.subList(startIndex, endIndex + 1)
                .stream()
                .flatMap(Set::stream)
                .collect(Collectors.toSet());
    }

    public static class Builder {
        private final List<Set<RzKey>> parts = new LinkedList<>();
        private int length = 1;

        public Builder addPart(Collection<RzKey> rzKeys) {
            parts.add(Set.copyOf(rzKeys));
            return this;
        }

        public Builder addPart(RzKey... rzKeys) {
            return addPart(Set.of(rzKeys));
        }

        public Builder withLength(int length) {
            this.length = length;
            return this;
        }

        public ProgressiveRzKeySelector build() {
            return new ProgressiveRzKeySelector(parts, length);
        }
    }
}

File: src/razer-sdk/sdk/ProgressiveRzKeySelector.ts
import {RzKey} from './RzKey.js';

export class ProgressiveRzKeySelector {
    private readonly length: number;
    private readonly parts: ReadonlySet<RzKey>[];
    private currentStep = 0;

    constructor(parts: ReadonlySet<RzKey>[], length = 1) {
        if (!parts || parts.length === 0) {
            throw new Error('There must be at least one part');
        }
        if (length <= 0) {
            throw new Error('Length must be a positive number');
        }
        if (length > parts.length) {
            throw new Error('Length cannot be larger than number of parts');
        }
        this.parts = parts;
        this.length = length;
    }

    getNextPart(): Set<RzKey> {
        const span = this.length - 1;
        if (this.currentStep === this.parts.length + span) {
            this.currentStep = 0;
        }

        let returnedPart: Set<RzKey>;
        if (this.currentStep - span <= 0) {
            returnedPart = this.joinSets(0, this.currentStep);
        } else if (this.currentStep >= this.parts.length - 1) {
            returnedPart = this.joinSets(this.currentStep - span, this.parts.length - 1);
        } else {
            returnedPart = this.joinSets(this.currentStep - span, this.currentStep);
        }

        this.currentStep++;
        return returnedPart;
    }

    getTotalSteps(): number {
        return this.parts.length + this.length - 1;
    }

    private joinSets(startIndex: number, endIndex: number): Set<RzKey> {
        const result = new Set<RzKey>();
        for (let i = startIndex; i <= endIndex; i++) {
            for (const key of this.parts[i]) {
                result.add(key);
            }
        }
        return result;
    }
}

export class ProgressiveRzKeySelectorBuilder {
    private readonly parts: Set<RzKey>[] = [];
    private length = 1;

    addPart(rzKeys: Iterable<RzKey>): this {
        this.parts.push(new Set(rzKeys));
        return this;
    }

    addPartFromKeys(...rzKeys: RzKey[]): this {
        return this.addPart(rzKeys);
    }

    withLength(length: number): this {
        this.length = length;
        return this;
    }

    build(): ProgressiveRzKeySelector {
        return new ProgressiveRzKeySelector(this.parts, this.length);
    }
}
